url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/PoulidorDistanceFraudSimulator.java

```diff
@@ -0,0 +1,189 @@
+			//System.out.println(mostFrequentSequence.toString());
+			result += (double)mostFrequentSequence.getTotalFrequency()/Math.pow(2, n);
+		}
+		return result/RUNS;
+	}
+	
+	private SequenceInGraph runForMostFrequentSequence(int n){
+		TreeMap<SequenceInGraph, SequenceInGraph> sequences = new TreeMap<>();
+		SequenceInGraph first = new SequenceInGraph("", -1, 1);
+		sequences.put(first, first);
+		boolean[] graph = generateRandomGraph(n);
+		//boolean[] graph = new boolean[]{false, true, false, false};
+		//System.out.println("Graph: "+ToStringUtility.toStringBooleanArray(graph));
+		sequences = runForMostFrequentSequence(n, sequences, graph);
+		//System.out.println("Frequency: "+sequences.firstKey().getTotalFrequency());
+		return sequences.firstKey();
+	}
+	
+	/*Trujillo- Mar 11, 2014
+	 * In this method, we have a list of sequences and their number of appearances. 
+	 * Different to the tree, in this method we keep the positions (the nodes) in the tree of
+	 * the sequences. Together with the position, we put the number of walks that generate 
+	 * this sequence until this position. 
+	 * So, let "s" a sequence and p1, ..., pm and f1 ... fm the positions and sequences
+	 * respectively for s. Each position pi generates two new positions pi+1 and pi+2 and
+	 * also two new (potentially equal) sequences. The frequency of both sequences is fi
+	 * regardless whether they are equal or not.
+	 * Frequencies are merged if a sequence "s" at some position "p" already exists
+	 * in the table. In this case both frequencies are summed. 
+	 * 
+	 * We always keep the table with a size lower than SIZE.
+	 * */
+	
+	/*Trujillo- Mar 11, 2014
+	 * TODO: Change the tree map by a sorted list using the class
+	 * SequenceInTree.*/
+	private TreeMap<SequenceInGraph, SequenceInGraph> runForMostFrequentSequence(int n, 
+			TreeMap<SequenceInGraph, SequenceInGraph> sequences, boolean[] graph){
+		if (n == 0) return sequences;
+		sequences = runForMostFrequentSequence(n-1, sequences, graph);
+		TreeMap<SequenceInGraph, SequenceInGraph> result = new TreeMap<>();
+		//we first develop all the current sequence a bit more
+		for (SequenceInGraph s : sequences.keySet()){
+			for (int pos : s.frequencies.keySet()){
+				long frequency = s.frequencies.get(pos);
+				int pos0 = pos+1;
+				int pos1 = pos+2;
+				String s0 = (graph[pos0])?s.sequence+"1":s.sequence+"0";
+				String s1 = (graph[pos1])?s.sequence+"1":s.sequence+"0";
+				SequenceInGraph s0In = result.get(new SequenceInGraph(s0));
+				if (s0In == null) {
+					SequenceInGraph tmp = new SequenceInGraph(s0, pos0, frequency);
+					if (result.size() > size){
+						SequenceInGraph last = result.lastKey();
+						if (tmp.getTotalFrequency() > last.getTotalFrequency()){
+							result.remove(last);
+							result.put(tmp, tmp);
+						}
+					}
+					else result.put(tmp, tmp);
+				}
+				else {
+					s0In.addFrequencyTo(pos1, frequency);
+					result.put(s0In, s0In);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 77:91, 93:107.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.